### PR TITLE
Add lister generator and refactor common code

### PR DESCRIPTION
**Pull request by fabianvf**
Signed-off-by: Nick Hale <njohnhale@gmail.com>

Rebase of #13

### CODE REVIEW

File: pkg/generators/clientgen/markers.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clientgen
+
+import "sigs.k8s.io/controller-tools/pkg/markers"
+
+var (
```
Comment (stevekuznetsov) [inline pkg/generators/clientgen/markers.go:21]: It looks like only two of these end up being reused, why do we need to pull out and export all of them?

File: pkg/internal/listergen/parser.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"text/template"
+
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+// funcMap contains the list of functions which are to be registered with
+// the templates.
+var funcMap = template.FuncMap{
+	"lowerFirst": func(s string) string {
```
Comment (njhale) [inline pkg/internal/listergen/parser.go:33]: nit: this is effectively `camelCase`, right?

File: examples/pkg/listers/example/v1/clustertesttype.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Code generated by kcp code-generator. DO NOT EDIT.
+
+package v1
+
+import (
+	apimachinerycache "github.com/kcp-dev/apimachinery/pkg/cache"
```
[the remark this answers is not in the document]
Comment (fabianvf) [reply examples/pkg/listers/example/v1/clustertesttype.go:25]: `kcp-dev/apimachinery` supersedes that package, the long-term goal is that this repo does not need to depend on a fork of kubernetes (:crossed_fingers: )

File: pkg/generators/clientgen/markers.go (hunk 1 above)
Comment (fabianvf) [reply pkg/generators/clientgen/markers.go:21]: I can move the others back to private

**Review by njhale — APPROVED** [review]
/lgtm

This is functional now, so I'd like to merge and follow up with refactoring afterwards.

**Review by stevekuznetsov — APPROVED** [review]

**Pull request merged**